python/cudf/cudf/tests/test_pack.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import pandas as pd
+
+from cudf._lib.copying import pack, unpack
+from cudf.core import DataFrame, GenericIndex
+from cudf.tests.utils import assert_eq
+
+
+def check_packed_equality(df):
+    # basic
+    assert_packed_frame_equality(df)
+    # sliced
+    assert_packed_frame_equality(df[:-1])
+    assert_packed_frame_equality(df[1:])
+    assert_packed_frame_equality(df[2:-2])
+    # sorted
+    sortvaldf = df.sort_values("vals")
+    assert isinstance(sortvaldf.index, GenericIndex)
+    assert_packed_frame_equality(sortvaldf)
+
+
+def assert_packed_frame_equality(df):
+    pdf = df.to_pandas()
+
+    packed = pack(df)
+    del df
+    unpacked = unpack(packed)
+
+    assert_eq(unpacked, pdf)
+
+
+def test_packed_dataframe_equality_numeric():
+    np.random.seed(0)
+    df = DataFrame()
+    nelem = 10
+    df["keys"] = np.arange(nelem, dtype=np.float64)
+    df["vals"] = np.random.random(nelem)
+
+    check_packed_equality(df)
+
+
+def test_packed_dataframe_equality_categorical():
+    np.random.seed(0)
+
+    df = DataFrame()
+    df["keys"] = pd.Categorical("aaabababac")
```

Comment:
This test, along with its pickled variant, fails because the unpacked DataFrame gives this column a dtype of `uint8` instead of `category`. Is there a way to support categorical columns here or should we just skip this test for now?

Reply:
https://github.com/rapidsai/cudf/blob/a600005493d32cd7049ddcc8816fd8885a4c45e9/python/cudf/cudf/core/frame.py#L2378

This can be of help I suppose? cc @shwina

Reply:
Yeah I chatted about this earlier, I think the best way forward here would be to include the columns' respective dtypes in the `PackedColumns` object so that they can be reapplied at unpack time.